### DISCUSSION TITLE
Getting react child error when passing JSX to the parameters

### DISCUSSION THREAD

**Discussion by NishargShah** [storybookjs/storybook#13922] [Help]
**Describe the bug**

Hello, I recently started to make documentation of my package but I got one stupid error while making my own addon that makes me sick. this addon just shows the static code of every story for copy-paste.

In `addon.js`, if I return the `example` variable its works fine and gives me the results in the `code` panel but if I passed `code` variable then it gives me the error, why??

**Error**

> Objects are not valid as a React child (found: object with keys {$$typeof, type, key, ref, props, _owner, _store}). If you meant to render a collection of children, use an array instead.

**Code snippets**

**register.js** ( already added in the `main.js` addons array )

```
import React from 'react';
import { AddonPanel } from '@storybook/components';
import { addons, types } from '@storybook/addons';
import Content from './addon';

addons.register('my/code-addon', () => {
  addons.add('code-addon/panel', {
    title: 'Code',
    type: types.PANEL,
    render: ({ active, key }) => (
      <AddonPanel active={active} key={key}>
        <Content />
      </AddonPanel>
    ),
  });
});
```

**addon.js**

```
import React from 'react';
import { useParameter } from '@storybook/api';

const Content = () => {
  const example = (
    <div>
      <p>Hello</p>
      <p>World</p>
    </div>
  );
  console.log(example);
  const code = useParameter('code', null);
  console.log('code', code);

  if (!code) return null;

  return code; // return example;
};

export default Content;
```

**Simple.stories.js**

```
import React from 'react';
import Mapper from './components/Mapper';

const SimpleApp = {
  title: 'Examples/Simple',
  component: Mapper,
  parameters: {
    controls: { hideNoControlsWarning: true },
    code: (
      <div>
        <p>Hello</p>
        <p>External</p>
      </div>
    ),
  },
};

export const Simple = () => <Mapper />;

Simple.args = {};

export default SimpleApp;
```

**Screenshots**
![image](https://user-images.githubusercontent.com/35843261/107259878-8e801f80-6a63-11eb-99e1-09e653f14352.png)

**System**

```
Ubuntu 20.04

@storybook/addon-actions: "^6.1.17",
@storybook/addon-essentials: "^6.1.17",
@storybook/addon-links: "^6.1.17",
@storybook/addon-storysource: "^6.1.17",
@storybook/addons: "^6.1.17",
@storybook/api: "^6.1.17",
@storybook/components: "^6.1.17",
@storybook/preset-scss: "^1.0.3",
@storybook/react: "^6.1.17",
@storybook/router: "^6.1.17",
@storybook/source-loader: "^6.1.17",
@storybook/theming: "^6.1.17",
```

**Comment by shilman** [comment]
Storybook is split between manager (Storybook UI code) and preview (user code). The two communicate via a channel, which serializes objects back and forth. See here: https://storybook.js.org/docs/react/addons/introduction#storybook-basics

In this case, your preview code is setting up the parameters as JSX, which gets compiled down to a non-serializable function. The manager code tries to read it, but it's getting whatever `telejson` serialized the parameter as.

To fix this you might need to serialize it yourself across the channel. Example: https://github.com/storybookjs/addon-kit/blob/main/src/withRoundTrip.js

**Reply by wagedomain** [reply]
I have a similar problem that NishargShah had. We're trying to pass an MDX object to an addon, the desired effect is to create a "design" set of docs in a tab sitting next to the developer Docs tab.

This worked wonderfully when everything was encapsulated inside the addon, but that's not scalable. Instead I want to essentially parameterize the story itself, so I can use the story Meta to link it to the design MDX file. I immediately ran into issues similar to this post.

Since this is a reusable addon, and we want it to be as simple as possible for our endusers to use, I'm trying to figure out the best way to structure this. How would you recommend passing in a component like the MDX file or some other component for distributed storybooks?

I was really hoping for something similar to:

```
import React from 'react';
import { Button } from 'component-library';
import { buttonDesignDoc } from './docs/design';

<Meta
  title: 'Components/Button',
  component: Button,
  parameters: {
    designDocs: buttonDesignDoc
  },
/>

// rest of story
```

**Reply by NishargShah** [reply]
I end up with passing the text to my `code` custom addon

look https://img-mapper.github.io/react-docs

**Reply by wagedomain** [reply]
Thanks for sending that! I looked at the code and I'm not sure where you're actually passing in `code`? I still see the register using `useParameter` but it's no longer being passed in?

I've had a very hard time serializing/deserializing an MDX component. I've followed this example: https://stackoverflow.com/questions/60112060/pass-a-jsx-element-to-storybook-parameters-in-a-custom-build-addon but my JSX element is still null, even though I can see an object being passed over.

**Reply by NishargShah** [reply]
as I said I am passing text on code and made custom addon by my self

code: https://github.com/img-mapper/react-docs/tree/master/.storybook/code-addon

don't forget to add register in the `main.js` https://github.com/img-mapper/react-docs/blob/master/.storybook/main.js

Here is how I am passing the text to the addon https://github.com/img-mapper/react-docs/blob/master/src/stories/Simple.stories.js#L25

Clone that repo and see how its working

**Reply by wagedomain** [reply]
Ah I see! Thank you for the clarification. 

I'm not sure this would work in my setup as-written, as I'm using MDX-based stories already. But it's something I can explore.

**Reply by AntonisFK** [reply]
@wagedomain running into a similar problem. What ended up being your solution?

**Reply by NishargShah** [reply]
You can check my above answer, I don't know how I resolved it before, haha
